Linkify usernames in ratings for admins

xlisachan opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/9297

---

@wagnerand , can you please let me know if I am on the right track with linkifying the usernames in the AddonReviewCard component?

<img width="873" alt="Screen Shot 2020-03-27 at 1 26 30 PM" src="https://user-images.githubusercontent.com/28129806/77784325-b70c5000-7030-11ea-8299-5f79e92bcbc6.png">

If so, can you provide some guidance? I added a variable (withAuthor) to determine whether `byLineString` should include a linkified author/username or just the author/username in plaintext. Initially I had the `urlStart`/`urlEnd` set to `a` tags, but realized in order to properly utilize `replaceStringsWithJSX` I would need to update them. I tried to follow how `replaceStringsWithJSX` was used in the `Errors - NotFound` component as guidance, but cannot seem to figure out what keeps breaking.

wagnerand commented:
@willdurand or @bobsilverberg could you help out here, please?

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -348,13 +348,10 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+      // translators: Example in English: "posted last week"
+      const byLineString = i18n.gettext(
+        'posted %(linkStart)s%(timestamp)s%(linkEnd)s',
+      );
```
So I think we want to keep the original `byLineString` because it is possible that translations are different in some languages (i.e. we cannot simply append both translation strings, we need the full sentence).

xlisachan replied:
Thanks for the insight! I have reverted `byLineString` back to its original format.

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -405,6 +402,43 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+        ],
+      });
+
+      const byAuthor = !review.userUrl ? byAuthorOnly : byAuthorLink;
```
The condition to render a link to the user profile is not about the `userUrl` in the `review` but whether the current user is an admin. Something like this: https://github.com/mozilla/addons-frontend/blob/48952905ceccb8e2209c76f063f8c649969fe388/src/amo/components/AddonAdminLinks/index.js#L159

xlisachan replied:
I had first created the condition with `userUrl` because I noticed that some `userUrl` were `null` in a handful of reviews. I thought this indicated that not all users had a profile page.

<img width="463" alt="Screen Shot 2020-03-30 at 4 02 02 PM" src="https://user-images.githubusercontent.com/28129806/77956439-df4db600-729f-11ea-9a8e-f6b91c369d65.png">

As advised, I used `hasAdminPermission` to check whether the current user is an admin.

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -405,6 +402,43 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+          <span>{byLineTimestamp}</span>
+        </>
+      );
+
```
I think we want something a bit different. I guess there should be 3 different translation strings for `byLineString`:

- when we have no review (current behavior)
- when we have a review and the current logged in user is not an admin (current behavior)
- when we have a review and the current logged in user is an admin (new!)

Then we'll have to update the code near `replaceStringsWithJSX()`, which is not very friendly. For each `<Link />`, this function needs:

- two placeholders that will be replaced by the same values (one for "start" and the other for "end"). I suggest we name these placeholders: `linkUserProfileStart` and `linkUserProfileEnd`. These placeholders are similar to `linkEnd` and `linkStart` in `byLineString`
- a new "replacement entry" in the `replacements` list (still in `replaceStringsWithJSX()` but a different key): This array is an array of arrays, and each array contains 3 parameters: the start placeholder, the end placeholder and a function.

I think the placeholders replaced by themselves should be `undefined` when there is no user OR the user is NOT an admin, too (so it is not just `linkUserProfileEnd: '%(linkUserProfileEnd)s'` but maybe something like `linkUserProfileEnd: !noAuthor && hasAdminPermission && '%(linkUserProfileEnd)s'` but that's just an example).

Next, we only want a new replacement entry if there is a user AND the user is an admin so we'll have to extract the `replacements` array in the `replaceStringsWithJSX()` call in a new variable (above the call to `replaceStringsWithJSX()`) and `.push()` a new replacement entry if the user is an admin.

After that, we can construct the `<Link />` in the function (third parameter of a replacement entry). This is straightforward because we only have to return a `<Link />` all the time, pointing to `/user/${review.userId}/`. We'll see about passing the `clientApp`/`lang` later (not sure that's needed).

Once we have that, we want to write new test cases (and make sure existing ones don't fail because they shouldn't).

xlisachan replied:
I did not realize that I could add conditions to the placeholders! This was extremely helpful, thank you!!!!  🎉

xlisachan commented:
Fixes #9297 

Thanks for the feedback, @willdurand ! When you have a moment, please review this pull request that linkifies usernames for admins.

As noted above, I have made the following changes:

- Reverted `byLineString`to its original format, and updated the conditional
- Added `hasAdminPermission` to check whether the current user is an admin
- Added the placeholders `linkUserProfileStart` and `linkUserProfileEnd`
- Extracted the `replacements` array and created a new variable to push the new replacement entry if the user is an admin
- Created the new replacement `<Link/>` entry for the username
- Created a test case to check that the username is linkified if the current user is an admin
- Created a test case to check that the username was not linkified if the current user is not an admin

I do not believe we need to pass `clientApp/lang`. The format of the link when clicked is as follows:
`http://localhost:3000/en-US/firefox/user/6506636/` - please confirm.

Also, as mentioned above, some of the users' `userUrl` have a value of `null`. I thought this meant the user did not have a profile page and would create a broken link if used. Can you confirm that `userUrl` can be left out of the condition?

Please let me know if you have any questions. Thanks!

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -371,38 +378,66 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
-          [
-            'linkStart',
-            'linkEnd',
+      const arrReplacements = [
```
I guess you can name this variable `replacements`

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -371,38 +378,66 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+        ],
+      ];
+
+      const getReplacements = () => {
```
Why would you need a function here?

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -371,38 +378,66 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+            'linkUserProfileEnd',
             (text) => {
-              return slugForReviewLink ? (
+              return !noAuthor && hasAdminPermission ? (
```
So I believe you should only push to `replacements` when both conditions are satisfied.

willdurand commented on src/amo/components/AddonReviewCard/index.js:
```diff
@@ -371,38 +378,66 @@ export class AddonReviewCardBase extends React.Component<InternalProps> {
+          linkEnd: '%(linkEnd)s',
+          linkStart: '%(linkStart)s',
+          linkUserProfileStart:
+            !noAuthor && hasAdminPermission ? '%(linkUserProfileStart)s' : null,
```
Shouldn't it be `undefined` instead of `null`?

I think the `!noAuthor && hasAdminPermission` check could be extracted in a new variable named `showUserProfileLink` or something.

willdurand commented:
> I do not believe we need to pass clientApp/lang. The format of the link when clicked is as follows: http://localhost:3000/en-US/firefox/user/6506636/ - please confirm.

Ok, that might be enough.

> Also, as mentioned above, some of the users' userUrl have a value of null. I thought this meant the user did not have a profile page and would create a broken link if used. Can you confirm that userUrl can be left out of the condition?

Yeah, I think this is a public property and not all users have a profile (but admins can still see some sort of profiles because they have elevated privileges). That is why we need to create the URL using the `userId`.

willdurand commented on tests/unit/amo/components/TestAddonReviewCard.js:
```diff
@@ -1189,6 +1189,28 @@ describe(__filename, () => {
+      expect(authorLinks.at(0)).toHaveProp('to', `/user/${review.userId}/`);
+    });
+
+    it('checks that username is not linkified if current user is not an admin', () => {
```
```suggestion
    it('does not linkify username if current user is not an admin', () => {
```

xlisachan commented:
@willdurand I have made the requested changes:
- Changed variable name from `arrReplacements` to `replacements`
- Revised getReplacements
- Added !noAuthor to the conditional
- Updated `null` to `undefined`
- Extracted `!noAuthor && hasAdminPermission` to a new variable - `showUserProfileLink `
- Revised test case

Please review and let me know if you have any additional feedback. Thank you!

xlisachan commented:
Ah, now I see why it didn't need to be a function! That makes sense, thanks!
